iroh-net/src/derp/http/client.rs

```diff
@@ -624,24 +640,44 @@ impl Client {
         if reg.nodes.is_empty() {
             return Err(ClientError::NoNodeForTarget(target));
         }
-        let mut first_err: Option<ClientError> = None;
-        // TODO (ramfox): these dials should probably happen in parallel, and we should return the
-        // first one to respond.
-        for node in reg.nodes.iter() {
-            if node.stun_only {
-                if first_err.is_none() {
-                    first_err = Some(ClientError::StunOnlyNodesFound(target.clone()));
+        // usually 1 IPv4, 1 IPv6 and 2x http
+        const DIAL_PARALLELISM: usize = 4;
+
+        let this = self.clone();
+        let mut dials = futures::stream::iter(reg.nodes.clone().into_iter())
+            .map(|node| {
+                let this = this.clone();
+                let target = target.clone();
+                async move {
+                    if node.stun_only {
+                        return Err(ClientError::StunOnlyNodesFound(target));
+                    }
+                    let conn = this.dial_node(&node).await;
+                    match conn {
+                        Ok(conn) => Ok((conn, node)),
+                        Err(e) => Err(e),
+                    }
+                }
+            })
+            .buffer_unordered(DIAL_PARALLELISM);
+
+        let mut first_err = None;
+        while let Some(res) = dials.next().await {
+            match res {
+                Ok((conn, node)) => {
+                    // return on the first successfull one
+                    trace!("dialed region");
+                    return Ok((conn, node));
+                }
+                Err(e) => {
+                    if first_err.is_none() {
+                        first_err = Some(e);
```

Comment:
related to @rklaehn 's comment below, unless you return a Vec of errors you'll always have an arbitrary choice.  Probably easier is to `tracing::error!()` log all the error cases here?  or at least `warn!`.

Reply:
fixed